extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluez/src/main/java/org/eclipse/smarthome/binding/bluetooth/bluez/handler/BlueZBridgeHandler.java

```diff
@@ -0,0 +1,267 @@
+        for (tinyb.BluetoothAdapter a : BluetoothManager.getBluetoothManager().getAdapters()) {
+            if (a.getAddress().equals(address.toString())) {
+                adapter = a;
+                updateStatus(ThingStatus.ONLINE);
+                if (discoveryActive) {
+                    adapter.startDiscovery();
+                }
+                discoveryJob = scheduler.scheduleWithFixedDelay(() -> {
+                    checkForNewDevices();
+                }, 0, 10, TimeUnit.SECONDS);
+                return;
+            }
+        }
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "No adapter for this address found.");
+    }
+
+    @Override
+    public ThingUID getUID() {
+        return getThing().getUID();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    @Override
+    public void addDiscoveryListener(BluetoothDiscoveryListener listener) {
+        discoveryListeners.add(listener);
+    }
+
+    @Override
+    public void removeDiscoveryListener(@Nullable BluetoothDiscoveryListener listener) {
+        discoveryListeners.remove(listener);
+    }
+
+    @Override
+    public void scanStart() {
+        adapter.setRssiDiscoveryFilter(-96);
+        adapter.startDiscovery();
+        for (tinyb.BluetoothDevice tinybDevice : adapter.getDevices()) {
+            synchronized (devices) {
+                logger.debug("Device {} has RSSI {}", tinybDevice.getAddress(), tinybDevice.getRSSI());
+                BluetoothDevice device = devices.get(tinybDevice.getAddress());
+                if (device == null) {
+                    createAndRegisterBlueZDevice(tinybDevice);
+                } else {
+                    // let's update the rssi and txpower values
+                    device.setRssi(tinybDevice.getRSSI());
+                    device.setTxPower(tinybDevice.getTxPower());
+                    // The Bluetooth discovery expects a complete list on every scan,
+                    // so we also have to report the already known devices.
+                    notifyDiscoveryListeners(device);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void scanStop() {
+        if (adapter.getDiscovering()) {
+            try {
+                adapter.stopDiscovery();
+            } catch (BluetoothException e) {
+                // tinyb often throws the exception
+                // tinyb.BluetoothException: GDBus.Error:org.bluez.Error.Failed: No discovery started
+                // here, although we first check whether discovery is active or not.
+                // As a workaround, we simply ignore this exception.
+            }
+        }
+    }
+
+    @Override
+    public BluetoothAddress getAddress() {
+        return address;
+    }
+
+    @Override
+    public BluetoothDevice getDevice(BluetoothAddress address) {
+        if (devices.containsKey(address.toString())) {
```

Comment:
No, since reads should never be blocked and writes need to check whether a previous write already created the same device.

Reply:
According to ReadWriteLock, the read lock may be held simultaneously by multiple reader threads, so long as there are no writers. The write lock is exclusive. It could be useful.